File: node-src/io/files.ts
import { readFile } from "node:fs/promises";

import { ProcessFailure } from "../domain/errors.js";

function decodeUtf8(content: Uint8Array, label: string): string {
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(content);
  } catch {
    throw new ProcessFailure({
      code: "INVALID_UTF8",
      category: "domain",
      message: `${label} is not valid UTF-8`,
    });
  }
}

export async function readRequiredUtf8(
  path: string,
  label: string,
): Promise<string> {
  try {
    return decodeUtf8(await readFile(path), label);
  } catch (error: unknown) {
    if (error instanceof ProcessFailure) {
      throw error;
    }
    throw new ProcessFailure({
      code: "READ_FAILED",
      category: "io",
      message: `unable to read ${label}`,
    });
  }
}

export async function readOptionalUtf8(
  path: string,
  label: string,
): Promise<string | null> {
  try {
    return decodeUtf8(await readFile(path), label);
  } catch (error: unknown) {
    if (
      typeof error === "object"
      && error !== null
      && "code" in error
      && error.code === "ENOENT"
    ) {
      return null;
    }
    if (error instanceof ProcessFailure) {
      throw error;
    }
    throw new ProcessFailure({
      code: "READ_FAILED",
      category: "io",
      message: `unable to read ${label}`,
    });
  }
}

File: go/internal/metadata/files.go
package metadata

import (
	"errors"
	"fmt"
	"os"
	"unicode/utf8"
)

// readOptionalUtf8 reads path as UTF-8 text, returning (nil, nil) when the
// file does not exist. Ports readOptionalUtf8 from the original implementation,
// kept package-private per this port's scope (this package does not create
// a separate io package): the original implementation's
// loadResourceMainOverride is the only caller this package ports, and it
// only ever calls the optional variant.
//
// Deliberately not ported: readRequiredUtf8 (the original implementation's other
// export, unused by anything in this package's scope) and the
// the original implementation ProcessFailure error model both functions
// raise there (code/category/retryable/details); nothing in this port
// inspects those fields, only the error's message, so plain Go errors
// carrying the same message text are a faithful substitute here.
func readOptionalUtf8(path, label string) (*string, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, fmt.Errorf("unable to read %s", label)
	}
	if !utf8.Valid(content) {
		return nil, fmt.Errorf("%s is not valid UTF-8", label)
	}
	text := string(content)
	return &text, nil
}

// ReadOptionalUTF8 exposes this package's readOptionalUtf8 (the port of
// readOptionalUtf8 in the original implementation) to the transform runner, which
// needs the identical absent-file-is-nil / other-error-fails contract for
// pull inputs and shared adoption-status sidecars. Same semantics, same
// error text; label feeds the error message exactly as in the Node source.
func ReadOptionalUTF8(path, label string) (*string, error) {
	return readOptionalUtf8(path, label)
}
